sympy/__init__.py

```diff
@@ -66,8 +66,10 @@ def __sympy_debug():
         Implies, Equivalent, ITE, POSform, SOPform, simplify_logic, bool_map,
         true, false, satisfiable)
 
-from .assumptions import (AppliedPredicate, Predicate, AssumptionsContext,
-        assuming, Q, ask, register_handler, remove_handler, refine)
+from .assumptions import (AskHandlerClass,
+        AppliedPredicate, Predicate, AssumptionsContext,
+        global_assumptions, assuming, Q, ask, register_handler, remove_handler,
+        refine)
```

Comment:
Do these things really need to be exposed at top-level? I can't imagine many users wanting to use these explicitly.

Reply:
We have been importing `register_handler`, `remove_handler` and `AssumptionContext`. Then why not import `AskHandlerClass` and `global_assumptions`?

Reply:
The `global_assumptions` object is best modified using `assuming` although I guess some users will want to just modify it at the top of a script. Personally I think that we should avoid the use of global assumptions and also assuming. It would be better to pass assumptions explicitly to any functions that can use them.

Either way I think that we have for too long added to many things to the top-level sympy namespace. It's difficult to change that now because once the name has been added it becomes a backward compatibility problem. That's why I ask about this now before more names are added there.

Reply:
Then I'll remove this change.

Reply:
What about removing `register_handler` and `remove_handler` from top level import? Since handler is changed people will encounter backwards compatibility issue anyway, and this will be the good chance to reduce the namespace.

Reply:
We should maintain backward compatibility unless there is a good reason to break it.

Why would anyone ever need to use `remove_handler`?

Looking at this I think that both the old and the new implementations seem quite convoluted. I'm pretty sure there's a much simpler way of doing all this.

Reply:
No one will need it unless they are using the old implementation. I will migrate the handlers to new design and deprecate the old one. It will be less complicated by then.

Reply:
The old design is extremely convoluted. It has taken some time for me to figure out how the handlers actually get attached to the predicates and the predicates get created and I really don't see why it needs to be so complicated.

I'm struggling to understand that and then why these changes are an improvement. They seem to make the situation more complicated.

Reply:
So, your opinion is that how handlers are registered needs to be improved, right?